README.md
# Priithon
Priithon is a platform for image analysis and image analysis algorithm development.

While the focus here is on algorithm development, it makes it very easy to develop applications that would not require any programming skills, i.e. making "end user" biologist's or astronomer's applications.

Priithon is easy to install: download a single zip archive and unpack. (It requires Python to be installed. More info is on the PriithonWiki.(FIXME LINK))

# Target audience
While Priithon is mainly used (by myself) as a daily work horse for looking at microscope images of some cell nuclei or other biology, there are many aspects that are not primarily geared for my own benefit:

- I want others to be able to install Priithon with minimal effort.

  This means:
  - Simple one-step unpacking of a package file
  - No system administrator permissions, or system knowledge, is required

- It is easy to make custom application, that allow someone to perform advanced image analysis tasks, who does not have any prior training in programming or scripting

- It is the best platform I know, to teach (complete) programming novices some basic skills in scripting
  - The Python programming language is generally considered clean and easy to learn
  - The PyShell interface gives helpful function descriptions and pop-up lists of available functions as you type

- Transparent modular multi-tier design going from interactive explorations to algorithms to scientist-oriented applications
 1. Priithon sessions are (auto-) saved into text files, documenting every step of interactive explorations
 2. Copy and paste into a Python-module file creates reusable functions
 3. Functions are then fine-tuned into cleanly parametrized algorithms
 4. Simple GUI interfaces visualize the set of "most-used" parameters and allow execution per mouse-click
 5. Organizing multiple GUI windows with few-line setup functions into a "drag-into-Priithon-to-execute" script file

__Code license__: _New BSD License_

__Labels__: _Python_, _numpy_, _SWIG_, _image analysis_, _multi dimensional_, _opengl_, _priithon_

# Documentation (old)
 - [PriithonHandbook](https://rawgit.com/sebhaase/priithon/master/priithon_docs/PriithonHandbook.html)
 - [PriithonTutorials](https://rawgit.com/sebhaase/priithon/master/priithon_docs/PriithonTutorials.html)
 - [PriithonModules epydoc](https://rawgit.com/sebhaase/priithon/master/priithon_docs/PriithonEpydoc/index.html)

# priithon
Automatically exported from code.google.com/p/priithon


ChangeLog.md
# Changelog #

### 2012-08-12 ###
  * added `Y.GuiParams` - `_setValueWithoutTriggerOrGUIupdate(n,v)`

### 2012-06-21 ###
  * changed `Y.vLeftViewSubRegion` - removed `color=False` option
  * added `Y.vLeftHorizProfile()`
  * added `Y.vLeftVertProfile()`
  * added `Y.vLeftZProfile()`
  * added `Y._vLeft_removeGfx()`
  * added `Y._vLeft_normalize_v_ids()`
  * added "on left click ..." popup menu to color viewer

### 2012-06-14 ###
  * added `PriConfig.shellCommandTimer`
  * CHANGED `colorAxis=='smart'` to consider ALL axes (also the last 2 ones)
### 2012-05-04 ###
  * changed (fixed) buttonbox center aligment of statictext,ceckbox,..
  * changed (fixed) buttonbox background color

### 2012-04-26 ###
  * added 'options...' to `U.saveImg(...)`

### 2012-01-31 ###
  * `Y.plotFit...()` understands non-int dataset argument to mean an array used for fitting instead datapoints found in plot

### 2012-01-25 ###
  * `guiParams.py::_holdParamEvents` now understands lists of param names
  * added  `guiParams.py::_enableParamGUIs(self, n=None, enable=True)`

### 2011-09-02 ###
  * added `PriConfig.viewerOriginLeftBottomDefault = 1`
  * added to `appendFilename=False)` option to `Y.vTitleSet()`
  * changed frame titles to indicate viewer orientation:
    * `id)` leftBottom;`id]` leftTop;`id>` FFT;`id}` rFFT

  * added `includeId=False` option to `Y.vTitleGet()`
  * fixed `viewerCommon.readGLviewport()` for `clip=True` and `originLeftBottom=0`

### 2011-09-01 ###
  * viewer/viewer2 shows `<filename>` for all data having `meta.filename`
  * removed `originLeftBottom` from `GLViewerCommon.__init__()` -- was unused
  * changed `originLeftBottom` default from `None` to `1` in
`GLViewer2.__init__()` -- `None` is invalid value here

  * added `meta.originLeftBottom` to loaded image file data
  * viewer and viewer2 now show TIFF,PNG,JPG,... correctly oriented !!

### 2011-08-31 ###
  * added "Clear import-completion cache" to `Options` menu

### 2011-08-26 ###
  * shell string popup options for directory, and print if file doesn't exist
  * added to `buttonbox` textcontrol `wx.TE_PASSWORD` support on "" or label trailing with '\\b'
  * added `guiParams::_bboxNewline()`
  * added `textMultiline=False` option to `guiParams::_bboxText()`
  * fixed  `guiParams` to set value even if text fields has (or just got to) length 0
  * added `Y.email()`

### 2011-08-19 ###
  * added TAB completion for `import` statements

### 2011-08-16 ###
  * added `execModule=None` option to `Y.guiParams::_guiBox`

### 2011-08-08 ###
  * added "p" type to `buttonBox` for `wx.Panel`
  * added `guiParams::_bboxPanel()`
  * added `panel=None` option to `Y.view()` and `Y.view2()`

  * `buttonbox`: only call initial `startNewRow()` if not done by user
  * added `panel` option to `Y.plotFigure()`
  * removed scipy `wxplt::plot_frame::__getattr__()` redirection-to-client-HACK
  * removed scipy `wxplt::plot_frame::update()`

### 2011-08-05 ###
  * added `shellDisplayHookStringPopup = True` to `PriConfig.py`
  * added `Y._guiDisplStringChoice()`

### 2011-07-20 ###
  * added option `secMax=100000` to `U.loadImg_iterSec`
  * added `meta` attribte for resulting array of `Y.loadImg_seq()` -- todo: return `mockNDarray`

### 2011-07-12 ###
  * `Y.DropFrame` now accepts `execStr` to be a callable -- todo: should be renamed...

### 2011-07-05 ###
  * added `panel=None` option to `Y.DropFrame`
  * added `_getParamsGUIsTopLevelParents()` to `Y.guiParams()`

### 2011-06-21 ###
  * changed `Y.DropFrame`: added `callAfter=False` option (trying to fix "hanging/busy" mouse pointer on Windows, when execStr takes long time)

### 2011-05-05 ###
  * added context help via `F1` key
  * NOTE: `wxHtml` has problems with `<`(less than) inside `<PRE>` -- so `>` is shown instead -- BAD HACK !!!
### 2011-04-29 ###
  * added `PriConfig.shellEditEmacs`

### 2011-04-19 ###
  * added `external=''` option to `Y.editor`
  * added shell popup for modules and functions
  * added `PriConfig.shellDisplayHookEditPopup=True`

### 2011-02-22 ###
  * changed `Y.guiParams` `_bboxFloat` and `_bboxInt` default argument `sliderWidth=100` (old: `sliderWidth=-1`)
  * refactored `guiParams.py`: standard functionality does not depend on `execModule`; now default '''changed''' for `execModule` to be `__main__` again, like in plain buttonboxes

### 2011-02-21 ###
  * added filename TAB completion to Priithon shell (PyShell,PyCrust)
  * removed `__main__.shell` -- use `sys.app.frame.shell` instead

### 2011-02-08 ###
  * added `remove dataset` sub-menu to plot window context menu

### 2011-02-03 ###
  * added key shortcut for turning "commandwise autosave" on/off

### 2011-01-25 ###
  * changed `guiParams._bboxText()` default to `textWeight=1`

### 2011-01-24 ###
  * added `U.nd__center_of_geometry()`

### 2011-01-10 ###
  * changed viewer shortcut `a` now calls `vAutoSizeFrame()`
    * `p` now to toggles between `phase` and `abs` view for complex data
  * added `Y.vAutoSizeFrame()`
  * added viewer context menu item for `Y.vAutoSizeFrame()`
  * fixed viewer centering for FFT and rFFT views

### 2011-01-05 ###
  * `all.py` now sets matplotlib backend to `wxagg` before `from matplotlib import pyplot as P`

### 2010-12-06 ###
  * added `PriConfig.viewerShowComplexVals = True`
  * added `Show Complex Vals (real,imag)` to `Options` menu

### 2010-11-29 ###
  * added `readerSDT` for reading Becker & Hickl SDT files

### 2010-11-18 ###
  * added `close all viewers` to `Priithon` menu
  * added `Y.vGammaGUI(vid=-1, gamma=.3)`
  * added `C` (capital) shortcut for gammaGUI in (single color) viewer
  * changed `guiParams` to ignore exceptions when converting string to value
  * renamed `PriConfig.displayHookNdarrayPopup` to `PriConfig.shellDisplayHookNdarrayPopup`
  * added `PriConfig.shellDefaultActionOnArray`
  * renamed arguments of `Y.vd()`, `Y.vTransferFct()`, `Y.vClose()` `id` to `vid`
  * added `verbose=True` option to `Y.saveSession()`
  * added `autosave()` to `py/shell.py`

PIL/ImageFile.py
#
# The Python Imaging Library.
# $Id: ImageFile.py 2930 2006-12-02 13:50:40Z fredrik $
#
# base class for image file handlers
#
# history:
# 1995-09-09 fl   Created
# 1996-03-11 fl   Fixed load mechanism.
# 1996-04-15 fl   Added pcx/xbm decoders.
# 1996-04-30 fl   Added encoders.
# 1996-12-14 fl   Added load helpers
# 1997-01-11 fl   Use encode_to_file where possible
# 1997-08-27 fl   Flush output in _save
# 1998-03-05 fl   Use memory mapping for some modes
# 1999-02-04 fl   Use memory mapping also for "I;16" and "I;16B"
# 1999-05-31 fl   Added image parser
# 2000-10-12 fl   Set readonly flag on memory-mapped images
# 2002-03-20 fl   Use better messages for common decoder errors
# 2003-04-21 fl   Fall back on mmap/map_buffer if map is not available
# 2003-10-30 fl   Added StubImageFile class
# 2004-02-25 fl   Made incremental parser more robust
#
# Copyright (c) 1997-2004 by Secret Labs AB
# Copyright (c) 1995-2004 by Fredrik Lundh
#
# See the README file for information on usage and redistribution.
#

import Image
import traceback, sys, string, os

MAXBLOCK = 65536

SAFEBLOCK = 1024*1024

ERRORS = {
    -1: "image buffer overrun error",
    -2: "decoding error",
    -3: "unknown error",
    -8: "bad configuration",
    -9: "out of memory error"
}

#
# --------------------------------------------------------------------
# Helpers

def _tilesort(t1, t2):
    # sort on offset
    return cmp(t1[2], t2[2])

#
# --------------------------------------------------------------------
# ImageFile base class

##
# Base class for image file handlers.

class ImageFile(Image.Image):
    "Base class for image file format handlers."

    def __init__(self, fp=None, filename=None):
        Image.Image.__init__(self)

        self.tile = None
        self.readonly = 1 # until we know better

        self.decoderconfig = ()
        self.decodermaxblock = MAXBLOCK

        if Image.isStringType(fp):
            # filename
            self.fp = open(fp, "rb")
            self.filename = fp
        else:
            # stream
            self.fp = fp
            self.filename = filename

        try:
            self._open()
        except IndexError, v: # end of data
            if Image.DEBUG > 1:
                traceback.print_exc()
            raise SyntaxError, v
        except TypeError, v: # end of data (ord)
            if Image.DEBUG > 1:
                traceback.print_exc()
            raise SyntaxError, v
        except KeyError, v: # unsupported mode
            if Image.DEBUG > 1:
                traceback.print_exc()
            raise SyntaxError, v
        except EOFError, v: # got header but not the first frame
            if Image.DEBUG > 1:
                traceback.print_exc()
            raise SyntaxError, v

        if not self.mode or self.size[0] <= 0:
            raise SyntaxError, "not identified by this driver"

    def draft(self, mode, size):
        "Set draft mode"

        pass

    def verify(self):
        "Check file integrity"

        # raise exception if something's wrong.  must be called
        # directly after open, and closes file when finished.
        self.fp = None

    def load(self):
        "Load image data based on tile list"

        pixel = Image.Image.load(self)

        if self.tile is None:
            raise IOError("cannot load this image")
        if not self.tile:
            return pixel

        self.map = None

        readonly = 0

        if self.filename and len(self.tile) == 1:
            # try memory mapping
            d, e, o, a = self.tile[0]
            if d == "raw" and a[0] == self.mode and a[0] in Image._MAPMODES:
                try:
                    if hasattr(Image.core, "map"):
                        # use built-in mapper
                        self.map = Image.core.map(self.filename)
                        self.map.seek(o)
                        self.im = self.map.readimage(
                            self.mode, self.size, a[1], a[2]
                            )
                    else:
                        # use mmap, if possible
                        import mmap
                        file = open(self.filename, "r+")
                        size = os.path.getsize(self.filename)
                        # FIXME: on Unix, use PROT_READ etc
                        self.map = mmap.mmap(file.fileno(), size)
                        self.im = Image.core.map_buffer(
                            self.map, self.size, d, e, o, a
                            )
                    readonly = 1
                except (AttributeError, EnvironmentError, ImportError):
                    self.map = None

        self.load_prepare()

        # look for read/seek overrides
        try:
            read = self.load_read
        except AttributeError:
            read = self.fp.read

        try:
            seek = self.load_seek
        except AttributeError:
            seek = self.fp.seek

        if not self.map:

            # sort tiles in file order
            self.tile.sort(_tilesort)

            try:
                # FIXME: This is a hack to handle TIFF's JpegTables tag.
                prefix = self.tile_prefix
            except AttributeError:
                prefix = ""

            for d, e, o, a in self.tile:
                d = Image._getdecoder(self.mode, d, a, self.decoderconfig)
                seek(o)
                try:
                    d.setimage(self.im, e)
                except ValueError:
                    continue
                b = prefix
                t = len(b)
                while 1:
                    s = read(self.decodermaxblock)
                    if not s:
                        #20080528 - seb HACK - CHECK -- why is this set to [] -- prevents debugging !?  -- self.tile = []
                        raise IOError("image file is truncated (%d bytes not processed)" % len(b))
                    b = b + s
                    n, e = d.decode(b)
                    if n < 0:
                        break
                    b = b[n:]
                    t = t + n

        self.tile = []
        self.readonly = readonly

        self.fp = None # might be shared

        if not self.map and e < 0:
            error = ERRORS.get(e, "decoder error %d" % e)
            raise IOError(error + " when reading image file")

        # post processing
        if hasattr(self, "tile_post_rotate"):
            # FIXME: This is a hack to handle rotated PCD's
            self.im = self.im.rotate(self.tile_post_rotate)
            self.size = self.im.size

        self.load_end()

        return Image.Image.load(self)

    def load_prepare(self):
        # create image memory if necessary
        if not self.im or\
           self.im.mode != self.mode or self.im.size != self.size:
            self.im = Image.core.new(self.mode, self.size)
        # create palette (optional)
        if self.mode == "P":
            Image.Image.load(self)

    def load_end(self):
        # may be overridden
        pass

    # may be defined for contained formats
    # def load_seek(self, pos):
    #     pass

    # may be defined for blocked formats (e.g. PNG)
    # def load_read(self, bytes):
    #     pass

##
# Base class for stub image loaders.
# <p>
# A stub loader is an image loader that can identify files of a
# certain format, but relies on external code to load the file.

class StubImageFile(ImageFile):
    "Base class for stub image loaders."

    def _open(self):
        raise NotImplementedError(
            "StubImageFile subclass must implement _open"
            )

    def load(self):
        loader = self._load()
        if loader is None:
            raise IOError("cannot find loader for this %s file" % self.format)
        image = loader.load(self)
        assert image is not None
        # become the other object (!)
        self.__class__ = image.__class__
        self.__dict__ = image.__dict__

    ##
    # (Hook) Find actual image loader.

    def _load(self):
        raise NotImplementedError(
            "StubImageFile subclass must implement _load"
            )

##
# (Internal) Support class for the <b>Parser</b> file.

class _ParserFile:
    # parser support class.

    def __init__(self, data):
        self.data = data
        self.offset = 0

    def close(self):
        self.data = self.offset = None

    def tell(self):
        return self.offset

    def seek(self, offset, whence=0):
        if whence == 0:
            self.offset = offset
        elif whence == 1:
            self.offset = self.offset + offset
        else:
            # force error in Image.open
            raise IOError("illegal argument to seek")

    def read(self, bytes=0):
        pos = self.offset
        if bytes:
            data = self.data[pos:pos+bytes]
        else:
            data = self.data[pos:]
        self.offset = pos + len(data)
        return data

    def readline(self):
        # FIXME: this is slow!
        s = ""
        while 1:
            c = self.read(1)
            if not c:
                break
            s = s + c
            if c == "\n":
                break
        return s

##
# Incremental image parser.  This class implements the standard
# feed/close consumer interface.

class Parser:

    incremental = None
    image = None
    data = None
    decoder = None
    finished = 0

    ##
    # (Consumer) Reset the parser.  Note that you can only call this
    # method immediately after you've created a parser; parser
    # instances cannot be reused.

    def reset(self):
        assert self.data is None, "cannot reuse parsers"

    ##
    # (Consumer) Feed data to the parser.
    #
    # @param data A string buffer.
    # @exception IOError If the parser failed to parse the image file.

    def feed(self, data):
        # collect data

        if self.finished:
            return

        if self.data is None:
            self.data = data
        else:
            self.data = self.data + data

        # parse what we have
        if self.decoder:

            if self.offset > 0:
                # skip header
                skip = min(len(self.data), self.offset)
                self.data = self.data[skip:]
                self.offset = self.offset - skip
                if self.offset > 0 or not self.data:
                    return

            n, e = self.decoder.decode(self.data)

            if n < 0:
                # end of stream
                self.data = None
                self.finished = 1
                if e < 0:
                    # decoding error
                    self.image = None
                    error = ERRORS.get(e, "decoder error %d" % e)
                    raise IOError(error + " when reading image file")
                else:
                    # end of image
                    return
            self.data = self.data[n:]

        elif self.image:

            # if we end up here with no decoder, this file cannot
            # be incrementally parsed.  wait until we've gotten all
            # available data
            pass

        else:

            # attempt to open this file
            try:
                try:
                    fp = _ParserFile(self.data)
                    im = Image.open(fp)
                finally:
                    fp.close() # explicitly close the virtual file
            except IOError:
                pass # not enough data
            else:
                flag = hasattr(im, "load_seek") or hasattr(im, "load_read")
                if flag or len(im.tile) != 1:
                    # custom load code, or multiple tiles
                    self.decode = None
                else:
                    # initialize decoder
                    im.load_prepare()
                    d, e, o, a = im.tile[0]
                    im.tile = []
                    self.decoder = Image._getdecoder(
                        im.mode, d, a, im.decoderconfig
                        )
                    self.decoder.setimage(im.im, e)

                    # calculate decoder offset
                    self.offset = o
                    if self.offset <= len(self.data):
                        self.data = self.data[self.offset:]
                        self.offset = 0

                self.image = im

    ##
    # (Consumer) Close the stream.
    #
    # @return An image object.
    # @exception IOError If the parser failed to parse the image file.

    def close(self):
        # finish decoding
        if self.decoder:
            # get rid of what's left in the buffers
            self.feed("")
            self.data = self.decoder = None
            if not self.finished:
                raise IOError("image was incomplete")
        if not self.image:
            raise IOError("cannot parse this image")
        if self.data:
            # incremental parsing not possible; reopen the file
            # not that we have all data
            try:
                fp = _ParserFile(self.data)
                self.image = Image.open(fp)
            finally:
                fp.close() # explicitly close the virtual file
        return self.image

# --------------------------------------------------------------------

##
# (Helper) Save image body to file.
#
# @param im Image object.
# @param fp File object.
# @param tile Tile list.

def _save(im, fp, tile):
    "Helper to save image based on tile list"

    im.load()
    if not hasattr(im, "encoderconfig"):
        im.encoderconfig = ()
    tile.sort(_tilesort)
    # FIXME: make MAXBLOCK a configuration parameter
    bufsize = max(MAXBLOCK, im.size[0] * 4) # see RawEncode.c
    try:
        fh = fp.fileno()
        fp.flush()
    except AttributeError:
        # compress to Python file-compatible object
        for e, b, o, a in tile:
            e = Image._getencoder(im.mode, e, a, im.encoderconfig)
            if o > 0:
                fp.seek(o, 0)
            e.setimage(im.im, b)
            while 1:
                l, s, d = e.encode(bufsize)
                fp.write(d)
                if s:
                    break
            if s < 0:
                raise IOError("encoder error %d when writing image file" % s)
    else:
        # slight speedup: compress to real file object
        for e, b, o, a in tile:
            e = Image._getencoder(im.mode, e, a, im.encoderconfig)
            if o > 0:
                fp.seek(o, 0)
            e.setimage(im.im, b)
            s = e.encode_to_file(fh, bufsize)
            if s < 0:
                raise IOError("encoder error %d when writing image file" % s)
    try:
        fp.flush()
    except: pass


##
# Reads large blocks in a safe way.  Unlike fp.read(n), this function
# doesn't trust the user.  If the requested size is larger than
# SAFEBLOCK, the file is read block by block.
#
# @param fp File handle.  Must implement a <b>read</b> method.
# @param size Number of bytes to read.
# @return A string containing up to <i>size</i> bytes of data.

def _safe_read(fp, size):
    if size <= 0:
        return ""
    if size <= SAFEBLOCK:
        return fp.read(size)
    data = []
    while size > 0:
        block = fp.read(min(size, SAFEBLOCK))
        if not block:
            break
        data.append(block)
        size = size - len(block)
    return string.join(data, "")
